util/src/test/java/io/grpc/util/AdvancedTlsX509TrustManagerTest.java

```diff
@@ -0,0 +1,147 @@
+  private File serverCert0File;
+
+  private X509Certificate[] caCert;
+  private X509Certificate[] serverCert0;
+
+  private ScheduledExecutorService executor;
+
+  @Before
+  public void setUp() throws IOException, GeneralSecurityException {
+    executor = new FakeClock().getScheduledExecutorService();
+    caCertFile = TestUtils.loadCert(CA_PEM_FILE);
+    caCert = CertificateUtils.getX509Certificates(TlsTesting.loadCert(CA_PEM_FILE));
+    serverCert0File = TestUtils.loadCert(SERVER_0_PEM_FILE);
+    serverCert0 = CertificateUtils.getX509Certificates(TlsTesting.loadCert(SERVER_0_PEM_FILE));
+  }
+
+  @Test
+  public void credentialSetting() throws Exception {
+    // Overall happy path checking of public API.
+    AdvancedTlsX509TrustManager trustManager = AdvancedTlsX509TrustManager.newBuilder().build();
+    trustManager.updateTrustCredentialsFromFile(serverCert0File);
+    assertArrayEquals(serverCert0, trustManager.getAcceptedIssuers());
+
+    trustManager.updateTrustCredentials(caCert);
+    assertArrayEquals(caCert, trustManager.getAcceptedIssuers());
+
+    trustManager.updateTrustCredentialsFromFile(serverCert0File, 1, TimeUnit.MINUTES,
+        executor);
+    assertArrayEquals(serverCert0, trustManager.getAcceptedIssuers());
+  }
+
+  @Test
+  public void credentialSettingParameterValidity() throws Exception {
+    // Checking edge cases of public API parameter setting.
+    AdvancedTlsX509TrustManager trustManager = AdvancedTlsX509TrustManager.newBuilder().build();
+
+    NullPointerException npe = assertThrows(NullPointerException.class, () -> trustManager
+        .updateTrustCredentials(null));
+    assertEquals("trustCerts", npe.getMessage());
+
+    npe = assertThrows(NullPointerException.class, () -> trustManager
+        .updateTrustCredentialsFromFile(null));
+    assertEquals("trustCertFile", npe.getMessage());
+
+    npe = assertThrows(NullPointerException.class, () -> trustManager
+        .updateTrustCredentialsFromFile(null, 1, null, null));
+    assertEquals("trustCertFile", npe.getMessage());
+
+    npe = assertThrows(NullPointerException.class, () -> trustManager
+        .updateTrustCredentialsFromFile(caCertFile, 1, null, null));
+    assertEquals("unit", npe.getMessage());
+
+    npe = assertThrows(NullPointerException.class, () -> trustManager
+        .updateTrustCredentialsFromFile(caCertFile, 1, TimeUnit.MINUTES, null));
+    assertEquals("executor", npe.getMessage());
+
+    Logger log = Logger.getLogger(AdvancedTlsX509TrustManager.class.getName());
+    TestHandler handler = new TestHandler();
+    log.addHandler(handler);
+    log.setUseParentHandlers(false);
+    log.setLevel(Level.FINE);
+    trustManager.updateTrustCredentialsFromFile(serverCert0File, -1, TimeUnit.SECONDS,
+        executor);
+    log.removeHandler(handler);
+    for (LogRecord record : handler.getRecords()) {
+      if (record.getMessage().contains("Default value of ")) {
+        assertTrue(true);
+        return;
```

Comment:
Ewww... If you're going to return early, let's not have it be part of a test that does lots of other things.

You might find Guava's `Iterables.find()` useful to change the code structure.

Reply:
Is that ok to use Stream API instead? I changed to
`LogRecord logRecord = handler.getRecords().stream()
        .filter(record -> record.getMessage().contains("Default value of "))
        .findFirst()
        .orElseThrow(() -> new AssertionError("Log message related to setting default "
            + "values not found"));
    assertNotNull(logRecord);`

Reply:
> The name here is pretty weak. updateTrustCredentials_replacesIssuers would seem better. 

Renamed.

> Otherwise, I'd have expected three separate instances of AdvancedTlsX509TrustManager to be used (and three separate tests then seems fair).

 My thinking process for this test is actually to use the same TM instance to show that it's fine to call various methods sequentially to load different version of certs.

> Although it seems updateTrustCredentialsFromFile(File) should also be run again at the end if the point is to notice the replacement. 

I didn't really get it - the test loads serverCert0, then caCert, and then serverCert0File. 3 total updates, 3 total `assertArrayEquals` - am I missing anything?